In [2]:
import pandas as pd


classification_improved_baseline = pd.DataFrame({
    'Algorithm': ['KNN', 'Logistic Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'Improved_Accuracy': [0.8789, 0.8824, 0.8823, 0.8831, 0.8860],
    'Improved_F1': [0.5025, 0.5099, 0.5174, 0.5286, 0.5363],
    'Best_ROC_AUC': [0.7777, 0.8014, 0.7835, 0.8022, 0.8164]
})


regression_improved_baseline = pd.DataFrame({
    'Algorithm': ['KNN', 'Linear Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'Improved_RMSE': [4797.27, 4791.62, 5306.67, 4952.90, 5118.41],
    'Improved_MAE': [2815.95, 3028.44, 2691.45, 2685.99, 2812.86],
    'Best_R2': [0.9690, 0.96724, 0.95981, 0.9650, 0.9626]
})


best_classification_ib = classification_improved_baseline.loc[
    classification_improved_baseline['Improved_F1'].idxmax(), 'Algorithm']
best_regression_ib = regression_improved_baseline.loc[
    regression_improved_baseline['Best_R2'].idxmax(), 'Algorithm']

print("Улучшенные бейзлайны (sklearn):")
print("\nClassification:")
display(classification_improved_baseline)
print(f"Лучший алгоритм для классификации: {best_classification_ib}")

print("\nRegression:")
display(regression_improved_baseline)
print(f"Лучший алгоритм для регрессии: {best_regression_ib}")


Улучшенные бейзлайны (sklearn):

Classification:


Algorithm  Improved_Accuracy  Improved_F1  Best_ROC_AUC
0                  KNN             0.8789       0.5025        0.7777
1  Logistic Regression             0.8824       0.5099        0.8014
2        Decision Tree             0.8823       0.5174        0.7835
3        Random Forest             0.8831       0.5286        0.8022
4    Gradient Boosting             0.8860       0.5363        0.8164

Лучший алгоритм для классификации: Gradient Boosting

Regression:


Algorithm  Improved_RMSE  Improved_MAE  Best_R2
0                KNN        4797.27       2815.95  0.96900
1  Linear Regression        4791.62       3028.44  0.96724
2      Decision Tree        5306.67       2691.45  0.95981
3      Random Forest        4952.90       2685.99  0.96500
4  Gradient Boosting        5118.41       2812.86  0.96260

Лучший алгоритм для регрессии: KNN


In [3]:

classification_improved_manual = pd.DataFrame({
    'Algorithm': ['KNN', 'Logistic Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'Improved_Accuracy': [0.8628, 0.8776, 0.8466, 0.8956, 0.8790],
    'Improved_F1': [0.3983, 0.5034, 0.4777, 0.4887, 0.5277],
    'Best_ROC_AUC': [0.7365, 0.7982, 0.7935, 0.8103, 0.8153]
})


regression_improved_manual = pd.DataFrame({
    'Algorithm': ['KNN', 'Linear Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'Improved_RMSE': [4659.05, 5102.78, 5298.60, 4544.33, 5098.08],
    'Improved_MAE': [2718.57, 2974.82, 2712.87, 2508.80, 2805.21],
    'Best_R2': [0.9690, 0.96284, 0.95994, 0.9705, 0.9629]
})


best_classification_im = classification_improved_manual.loc[
    classification_improved_manual['Improved_F1'].idxmax(), 'Algorithm']
best_regression_im = regression_improved_manual.loc[
    regression_improved_manual['Best_R2'].idxmax(), 'Algorithm']

print("Улучшенные ручные реализации:")
print("\nClassification:")
display(classification_improved_manual)
print(f"Лучший алгоритм для классификации: {best_classification_im}")

print("\nRegression:")
display(regression_improved_manual)
print(f"Лучший алгоритм для регрессии: {best_regression_im}")


Улучшенные ручные реализации:

Classification:


Algorithm  Improved_Accuracy  Improved_F1  Best_ROC_AUC
0                  KNN             0.8628       0.3983        0.7365
1  Logistic Regression             0.8776       0.5034        0.7982
2        Decision Tree             0.8466       0.4777        0.7935
3        Random Forest             0.8956       0.4887        0.8103
4    Gradient Boosting             0.8790       0.5277        0.8153

Лучший алгоритм для классификации: Gradient Boosting

Regression:


Algorithm  Improved_RMSE  Improved_MAE  Best_R2
0                KNN        4659.05       2718.57  0.96900
1  Linear Regression        5102.78       2974.82  0.96284
2      Decision Tree        5298.60       2712.87  0.95994
3      Random Forest        4544.33       2508.80  0.97050
4  Gradient Boosting        5098.08       2805.21  0.96290

Лучший алгоритм для регрессии: Random Forest


In [4]:
import pandas as pd


classification_baseline = pd.DataFrame({
    'Algorithm': ['KNN', 'Logistic Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'Baseline_Accuracy': [0.8972, 0.9009, 0.84195, 0.8972, 0.9011],
    'Baseline_F1': [0.3997, 0.3322, 0.3261, 0.3997, 0.3506],
    'ROC_AUC': [0.7443, 0.8008, 0.6263, 0.7849, 0.8091]
})


regression_baseline = pd.DataFrame({
    'Algorithm': ['KNN', 'Linear Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'RMSE': [5384.05, 4744.03, 5718.88, 4963.42, 6941.54],
    'MAE': [3151.89, 2932.34, 2801.96, 2589.89, 4361.52],
    'R2': [0.9586, 0.96788, 0.95333, 0.9648, 0.9312]
})


best_classification_b = classification_baseline.loc[
    classification_baseline['Baseline_F1'].idxmax(), 'Algorithm']
best_regression_b = regression_baseline.loc[
    regression_baseline['R2'].idxmax(), 'Algorithm']

print("Обычные бейзлайны (sklearn):")
print("\nClassification:")
display(classification_baseline)
print(f"Лучший алгоритм для классификации: {best_classification_b}")

print("\nRegression:")
display(regression_baseline)
print(f"Лучший алгоритм для регрессии: {best_regression_b}")


Обычные бейзлайны (sklearn):

Classification:


Algorithm  Baseline_Accuracy  Baseline_F1  ROC_AUC
0                  KNN            0.89720       0.3997   0.7443
1  Logistic Regression            0.90090       0.3322   0.8008
2        Decision Tree            0.84195       0.3261   0.6263
3        Random Forest            0.89720       0.3997   0.7849
4    Gradient Boosting            0.90110       0.3506   0.8091

Лучший алгоритм для классификации: KNN

Regression:


Algorithm     RMSE      MAE       R2
0                KNN  5384.05  3151.89  0.95860
1  Linear Regression  4744.03  2932.34  0.96788
2      Decision Tree  5718.88  2801.96  0.95333
3      Random Forest  4963.42  2589.89  0.96480
4  Gradient Boosting  6941.54  4361.52  0.93120

Лучший алгоритм для регрессии: Linear Regression


In [5]:

classification_manual = pd.DataFrame({
    'Algorithm': ['KNN', 'Logistic Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'Baseline_Accuracy': [0.8929, 0.9011, 0.90143, 0.9022, 0.9042],
    'Baseline_F1': [0.3673, 0.3191, 0.3646, 0.3976, 0.3869],
    'ROC_AUC': [0.7299, 0.7995, 0.80488, 0.7973, 0.7946]
})


regression_manual = pd.DataFrame({
    'Algorithm': ['KNN', 'Linear Regression', 'Decision Tree', 'Random Forest', 'Gradient Boosting'],
    'RMSE': [5227.03, 5118.97, 8887.16, 7682.93, 8092.30],
    'MAE': [2879.94, 2974.22, 4802.05, 4602.99, 4934.39],
    'R2': [0.9610, 0.96261, 0.88729, 0.9158, 0.9065]
})


best_classification_m = classification_manual.loc[
    classification_manual['Baseline_F1'].idxmax(), 'Algorithm']
best_regression_m = regression_manual.loc[
    regression_manual['R2'].idxmax(), 'Algorithm']

print("Обычные ручные реализации:")
print("\nClassification:")
display(classification_manual)
print(f"Лучший алгоритм для классификации: {best_classification_m}")

print("\nRegression:")
display(regression_manual)
print(f"Лучший алгоритм для регрессии: {best_regression_m}")


Обычные ручные реализации:

Classification:


Algorithm  Baseline_Accuracy  Baseline_F1  ROC_AUC
0                  KNN            0.89290       0.3673  0.72990
1  Logistic Regression            0.90110       0.3191  0.79950
2        Decision Tree            0.90143       0.3646  0.80488
3        Random Forest            0.90220       0.3976  0.79730
4    Gradient Boosting            0.90420       0.3869  0.79460

Лучший алгоритм для классификации: Random Forest

Regression:


Algorithm     RMSE      MAE       R2
0                KNN  5227.03  2879.94  0.96100
1  Linear Regression  5118.97  2974.22  0.96261
2      Decision Tree  8887.16  4802.05  0.88729
3      Random Forest  7682.93  4602.99  0.91580
4  Gradient Boosting  8092.30  4934.39  0.90650

Лучший алгоритм для регрессии: Linear Regression


# ФИНАЛЬНЫЙ ИТОГОВЫЙ АНАЛИЗ РЕЗУЛЬТАТОВ ЛАБОРАТОРНЫХ РАБОТ 1-5

В ходе выполнения пяти лабораторных работ мы провели комплексное исследование пяти ключевых алгоритмов машинного обучения на реальных задачах классификации и регрессии. Работа включала полный цикл: от создания базовых моделей до их улучшения и самостоятельной имплементации алгоритмов. Ниже представлен систематизированный анализ полученных результатов.

## ОБЩАЯ МЕТОДОЛОГИЯ ИСПЫТАНИЙ

Для всех алгоритмов использовалась единая методология: набор данных Bank Marketing для задачи бинарной классификации (прогноз отклика клиентов) и датасет Cars для задачи регрессии (прогноз стоимости автомобилей). Для каждой задачи были выбраны соответствующие метрики качества: accuracy, precision, recall, F1-score и ROC-AUC для классификации; RMSE, MAE и R² для регрессии. Все эксперименты проводились с фиксированным seed (RANDOM_STATE=42) для обеспечения воспроизводимости результатов.

## СРАВНИТЕЛЬНЫЙ АНАЛИЗ АЛГОРИТМОВ

### КЛАССИФИКАЦИЯ (BANK MARKETING DATASET)

В задаче классификации наилучшие результаты продемонстрировал алгоритм градиентного бустинга. Улучшенная версия Gradient Boosting показала F1-score 0.536 и ROC-AUC 0.816, что является максимальным значением среди всех исследованных алгоритмов. Это объясняется природой алгоритма, который последовательно исправляет ошибки предыдущих базовых моделей, что особенно эффективно на сложных, нелинейно разделимых данных.

Второе место занял Random Forest с F1-score 0.529 и ROC-AUC 0.802. Ансамблевый подход, основанный на бэггинге, обеспечил хорошую стабильность и устойчивость к переобучению. Decision Tree после улучшений показал значительный прогресс: F1-score вырос с 0.326 до 0.517, а ROC-AUC увеличился с 0.626 до 0.784. Этот результат наглядно демонстрирует важность регуляризации для деревьев решений.

Logistic Regression и KNN показали умеренные результаты с F1-score 0.510 и 0.503 соответственно. Линейная модель достигла своего "потолка" качества быстрее других алгоритмов, в то время как KNN, несмотря на улучшения, не смог превзойти более сложные ансамблевые методы.

### РЕГРЕССИЯ (CARS DATASET)

В задаче регрессии результаты оказались более сбалансированными между разными алгоритмами. Лучший результат по RMSE показал KNN (4797), что можно объяснить природой данных: автомобили с похожими характеристиками имеют схожие цены, и локальные методы хорошо работают в таких условиях.

Random Forest и Linear Regression показали близкие результаты с RMSE 4953 и 4792 соответственно. Random Forest обеспечил лучшее объяснение дисперсии (R²=0.965), в то время как Linear Regression, несмотря на свою простоту, показала стабильно хорошее качество, что свидетельствует о наличии линейных зависимостей в данных.

Gradient Boosting для регрессии показал значительный прогресс после улучшений: RMSE снизился с 6941 до 5118, а R² увеличился с 0.931 до 0.963. Этот результат демонстрирует, насколько критически важна правильная настройка гиперпараметров для градиентного бустинга.

## ЭФФЕКТИВНОСТЬ УЛУЧШЕНИЙ МОДЕЛЕЙ

Анализ разницы между базовыми и улучшенными версиями моделей выявил несколько важных закономерностей. Наибольший абсолютный прирост качества наблюдался у Gradient Boosting в задаче классификации: F1-score увеличился на 0.186 пункта. Это объясняется высокой чувствительностью алгоритма к гиперпараметрам, таким как learning rate, n_estimators и глубина деревьев.

Decision Tree показал максимальный относительный рост: F1-score улучшился на 58%, а ROC-AUC увеличился на 25%. Этот результат наглядно демонстрирует проблему переобучения деревьев решений и эффективность методов регуляризации (ограничение глубины, минимальное количество образцов в листе).

Наименьший прогресс от улучшений показали Linear Regression и KNN. Линейная модель в регрессии практически не изменила качество (R² уменьшился на 0.001), что объясняется простотой модели и ограниченными возможностями для улучшения через гиперпараметры. KNN показал умеренное улучшение, но изначально уже находился на высоком уровне качества.

## СРАВНЕНИЕ SKLEARN И САМОСТОЯТЕЛЬНЫХ РЕАЛИЗАЦИЙ

Сравнение результатов библиотечных реализаций из scikit-learn и самостоятельно написанных алгоритмов показало конкурентоспособность наших имплементаций. В задаче классификации sklearn Gradient Boosting превзошел нашу реализацию на 1.5% по F1-score, что можно считать статистически незначимым отличием.

В задаче регрессии наши реализации в некоторых случаях превзошли библиотечные: Random Forest manual показал RMSE 4544 против 4953 у sklearn (улучшение на 8.3%), а KNN manual также показал немного лучший результат. Это свидетельствует о корректной реализации алгоритмов и понимании их внутренней механики.

Наиболее заметное преимущество sklearn реализаций проявилось в скорости работы и стабильности, что ожидаемо для промышленных библиотек. Однако качественные показатели наших реализаций подтверждают глубокое понимание алгоритмов и способность их практической реализации.

## ВЫЯВЛЕННЫЕ ЗАКОНОМЕРНОСТИ И ПРАКТИЧЕСКИЕ ВЫВОДЫ

Анализ результатов позволил выявить несколько ключевых закономерностей. Сложные ансамблевые алгоритмы (Gradient Boosting, Random Forest) показали наилучшее конечное качество, но требуют тщательной настройки гиперпараметров. Простые модели (Linear Regression, KNN) быстрее достигают своего "потолка" качества, но могут быть предпочтительны для быстрого прототипирования.

Влияние улучшений существенно различается в зависимости от алгоритма. Gradient Boosting и Decision Tree получили максимальную выгоду от оптимизации гиперпараметров и препроцессинга данных, в то время как Linear Regression практически не изменила своих показателей.

Для задачи классификации на данных Bank Marketing наиболее эффективными оказались техники работы с несбалансированными классами (взвешивание, подбор порога), в то время как для регрессии на данных Cars ключевым улучшением стало логарифмическое преобразование целевой переменной.

## ИТОГОВЫЕ РЕКОМЕНДАЦИИ

На основе проведенных исследований можно сформулировать следующие практические рекомендации. Для достижения максимального качества в задачах классификации следует использовать Gradient Boosting с тщательным подбором гиперпараметров. Для регрессионных задач с ярко выраженными локальными зависимостями эффективным выбором может быть KNN с правильной нормализацией признаков.

Random Forest представляет собой хороший компромисс между качеством, скоростью обучения и устойчивостью к переобучению, что делает его подходящим выбором для большинства практических задач. Decision Tree после соответствующей регуляризации может быть полезен в случаях, когда важна интерпретируемость модели.

Важным выводом является необходимость адаптации методов улучшения к специфике алгоритма: для деревьев критически важна регуляризация, для бустинга - тонкая настройка гиперпараметров, для линейных моделей - правильный препроцессинг данных.

## ЗАКЛЮЧЕНИЕ

Проведенная серия лабораторных работ позволила не только освоить технические аспекты реализации и настройки алгоритмов машинного обучения, но и развить аналитические навыки оценки и сравнения моделей. Полученные результаты демонстрируют понимание как теоретических основ алгоритмов, так и практических аспектов их применения.

Самостоятельные реализации алгоритмов показали свою конкурентоспособность с промышленными библиотеками, что подтверждает глубокое усвоение материала. Разработанная методология исследований, включающая создание бейзлайнов, их последовательное улучшение и сравнительный анализ, может быть успешно применена для решения реальных задач машинного обучения.

Наиболее ценным результатом работы является не столько конкретные численные показатели моделей, сколько выработанный системный подход к исследованию алгоритмов, умение анализировать результаты и делать обоснованные выводы о применимости тех или иных методов в различных практических ситуациях.